src/hotspot/share/opto/predicates.cpp

```diff
@@ -147,3 +148,331 @@ Node* PredicateBlock::skip_regular_predicates(Node* regular_predicate_proj, Deop
+  }
+
+  // Pop a node from the stack and do the following with the node now being on top:
+  // - Is top node part of Template Assertion Predicate Bool?
+  //   - No? Done.
+  //   - Yes:
+  //     - Cloned before?
+  //       - No? Clone and replace node on top with cloned version.
+  //     - Rewire the node on top to the previous node on top (i.e. the cloned node to its new cloned/transformed
+  //       parent).
+  void pop_node() {
+    Node* previously_visited_parent = _stack.pop();
+    if (_stack.is_not_empty()) {
+      if (must_clone_node_on_top(previously_visited_parent)) {
+        clone_and_replace_top_node();
+        rewire_node_on_top_to(previously_visited_parent);
+      } else if (is_cloned_node(previously_visited_parent)) {
+        rewire_node_on_top_to(previously_visited_parent);
+      } // Else: Node is not part of the Template Assertion Predicate Bool (i.e. not on the chain to an OpaqueLoop* node)
+    }
+  }
+
+  bool is_cloned_node(Node* node) const {
+    return node->_idx >= _index_before_cloning;
+  }
+
+  // `new_parent` replaced the old parent of the current node on top of the stack for which we stored:
+  //     [top, i]
+  // where
+  //     top->in(i) = old_parent.
+  // We rewire the top node to the new parent such that:
+  //     top->in(i) = new_parent
+  void rewire_node_on_top_to(Node* new_parent) {
+    Node* top = _stack.top();
+    assert(is_cloned_node(top) && !top->is_Opaque1(), "must be cloned node on chain to OpaqueLoop*Node (excluded)");
+    const uint index_to_old_parent = _stack.node_index_to_previously_visited_parent();
+    top->set_req(index_to_old_parent, new_parent);
+  }
+
+ public:
+  CloneTemplateAssertionPredicateBool(BoolNode* template_bool, Node* ctrl_for_clones, PhaseIdealLoop* phase)
+      : _stack(template_bool),
+        _phase(phase),
+        _index_before_cloning(phase->C->unique()),
+        _ctrl_for_clones(ctrl_for_clones)
+  DEBUG_ONLY(COMMA _found_init(false)) {}
+
+  // Look for the OpaqueLoop* nodes to transform them with the strategy defined with 'transform_opaque_loop_nodes'.
+  // Clone all nodes in between.
+  BoolNode* clone(TransformOpaqueLoopNodes* transform_opaque_loop_nodes) {
+    Node* current;
+    while (_stack.is_not_empty()) {
+      current = _stack.top();
+      if (current->is_Opaque1()) {
+        transform_opaque_loop_node(current, transform_opaque_loop_nodes);
+        pop_transformed_opaque_loop_node();
+      } else if (!_stack.push_next_unvisited_input()) {
+        pop_node();
+      }
+    }
+    assert(current->is_Bool() && current->_idx >= _index_before_cloning, "new BoolNode expected");
+    assert(_found_init, "OpaqueLoopInitNode must always be found");
+    return current->as_Bool();
+  }
+};
+
+// This class caches a single OpaqueLoopInitNode and OpaqueLoopStrideNode. If the node is not cached, yet, we clone it
+// and store the clone in the cache to be returned for subsequent calls.
+class CachedOpaqueLoopNodes {
```

Comment:
Maybe it should say that it cashes the cloned opaque loop nodes in the name? Right now I would think you are caching the original loop nodes,  at least at first.
```suggestion
class ClonedOpaqueLoopNodesCache {
```